File: server/internal/validator/validator.go
package validator

import (
	"errors"
	"regexp"
	"strings"
	"unicode/utf8"
)

// Максимальные длины полей
const (
	MaxMessageLength     = 5000
	MaxNameLength        = 100
	MaxPhoneLength       = 30
	MaxAddressLength     = 200
	MaxCityLength        = 100
	MaxRegionLength      = 100
	MaxPostalCodeLength  = 20
	MaxCountryCodeLength = 2
	MaxWishlistLength    = 5000
	MaxAboutLength       = 5000
)

// Опасные паттерны для SQL/NoSQL injection
var dangerousPatterns = []*regexp.Regexp{
	// SQL injection
	regexp.MustCompile(`(?i)(union|select|insert|update|delete|drop|create|alter|exec|execute|script|javascript|<script)`),
	// XSS
	regexp.MustCompile(`(?i)(<script|<iframe|<object|<embed|<img.*onerror|javascript:)`),
	// Command injection (скобки ( ) разрешены)
	regexp.MustCompile(`(?i)(;.*\||\$\{|` + "`" + `)`),
}

// SanitizeString очищает строку от опасных символов
func SanitizeString(s string) string {
	// Убираем null bytes
	s = strings.ReplaceAll(s, "\x00", "")
	// Убираем лишние пробелы
	s = strings.TrimSpace(s)
	return s
}

// ValidateMessage проверяет сообщение чата
func ValidateMessage(content string) error {
	content = SanitizeString(content)

	if content == "" {
		return errors.New("message cannot be empty")
	}

	if utf8.RuneCountInString(content) > MaxMessageLength {
		return errors.New("message too long")
	}

	// Проверяем на опасные паттерны
	for _, pattern := range dangerousPatterns {
		if pattern.MatchString(content) {
			return errors.New("message contains prohibited content")
		}
	}

	return nil
}

// ValidateProfileField проверяет отдельное поле профиля
func ValidateProfileField(fieldName, value string, maxLength int) error {
	if value == "" {
		return nil // Пустые поля допустимы
	}

	value = SanitizeString(value)

	if utf8.RuneCountInString(value) > maxLength {
		return errors.New(fieldName + " is too long")
	}

	// Проверяем на опасные паттерны
	for _, pattern := range dangerousPatterns {
		if pattern.MatchString(value) {
			return errors.New(fieldName + " contains prohibited content")
		}
	}

	return nil
}

// ValidatePhone проверяет формат телефона
func ValidatePhone(phone string) error {
	if phone == "" {
		return nil
	}

	phone = SanitizeString(phone)

	if len(phone) > MaxPhoneLength {
		return errors.New("phone number is too long")
	}

	// Разрешаем только цифры, +, -, (, ), пробелы
	phoneRegex := regexp.MustCompile(`^[\d\s\+\-\(\)]+$`)
	if !phoneRegex.MatchString(phone) {
		return errors.New("phone number contains invalid characters")
	}

	return nil
}

// ValidateCountryCode проверяет код страны (ISO 3166-1 alpha-2)
func ValidateCountryCode(code string) error {
	if code == "" {
		return nil
	}

	code = SanitizeString(code)

	if len(code) != MaxCountryCodeLength {
		return errors.New("invalid country code")
	}

	// Только буквы A-Z
	codeRegex := regexp.MustCompile(`^[A-Za-z]{2}$`)
	if !codeRegex.MatchString(code) {
		return errors.New("country code must be 2 letters")
	}

	return nil
}

// ValidatePostalCode проверяет почтовый индекс
func ValidatePostalCode(code string) error {
	if code == "" {
		return nil
	}

	code = SanitizeString(code)

	if len(code) > MaxPostalCodeLength {
		return errors.New("postal code is too long")
	}

	// Разрешаем цифры, буквы, пробелы, дефисы
	postalRegex := regexp.MustCompile(`^[\d\w\s\-]+$`)
	if !postalRegex.MatchString(code) {
		return errors.New("postal code contains invalid characters")
	}

	return nil
}

// ValidateProfileData проверяет все данные профиля
func ValidateProfileData(data map[string]string) error {
	// Проверяем телефон
	if err := ValidatePhone(data["phone"]); err != nil {
		return err
	}

	// Проверяем страну
	if err := ValidateCountryCode(data["country"]); err != nil {
		return err
	}

	// Проверяем почтовый индекс
	if err := ValidatePostalCode(data["postal_code"]); err != nil {
		return err
	}

	// Проверяем текстовые поля
	fields := map[string]int{
		"about":            MaxAboutLength,
		"address_line1":    MaxAddressLength,
		"address_line2":    MaxAddressLength,
		"address_line1_en": MaxAddressLength,
		"address_line2_en": MaxAddressLength,
		"city":             MaxCityLength,
		"city_en":          MaxCityLength,
		"region":           MaxRegionLength,
		"region_en":        MaxRegionLength,
		"wishlist":         MaxWishlistLength,
		"anti_wishlist":    MaxWishlistLength,
	}

	for field, maxLen := range fields {
		if err := ValidateProfileField(field, data[field], maxLen); err != nil {
			return err
		}
	}

	return nil
}

File: client/src/utils/validator.ts
// Максимальные длины полей (соответствуют серверным)
export const MAX_MESSAGE_LENGTH = 10000;
export const MAX_PHONE_LENGTH = 30;
export const MAX_ADDRESS_LENGTH = 200;
export const MAX_CITY_LENGTH = 100;
export const MAX_REGION_LENGTH = 100;
export const MAX_POSTAL_CODE_LENGTH = 20;
export const MAX_WISHLIST_LENGTH = 10000;
export const MAX_ABOUT_LENGTH = 10000;

// Опасные паттерны для SQL/NoSQL injection и XSS
const DANGEROUS_PATTERNS = [
  // SQL injection
  /(union|select|insert|update|delete|drop|create|alter|exec|execute)\s/i,
  // XSS
  /(<script|<iframe|<object|<embed|javascript:)/i,
  /<img[^>]*onerror/i,
];

/**
 * Проверяет строку на наличие опасного содержимого
 */
export const containsDangerousContent = (value: string): boolean => {
  return DANGEROUS_PATTERNS.some((pattern) => pattern.test(value));
};

/**
 * Очищает строку от null bytes и лишних пробелов
 */
export const sanitizeString = (value: string): string => {
  return value.replace(/\x00/g, "").trim();
};

/**
 * Валидация сообщения чата
 */
export const validateMessage = (content: string): string | null => {
  const sanitized = sanitizeString(content);

  if (!sanitized) {
    return "Message cannot be empty";
  }

  if (sanitized.length > MAX_MESSAGE_LENGTH) {
    return `Message is too long (max ${MAX_MESSAGE_LENGTH} characters)`;
  }

  if (containsDangerousContent(sanitized)) {
    return "Message contains prohibited content";
  }

  return null;
};

/**
 * Валидация поля профиля
 */
export const validateProfileField = (
  value: string,
  maxLength: number,
  fieldName: string
): string | null => {
  if (!value) {
    return null; // Пустые поля допустимы
  }

  const sanitized = sanitizeString(value);

  if (sanitized.length > maxLength) {
    return `${fieldName} is too long (max ${maxLength} characters)`;
  }

  if (containsDangerousContent(sanitized)) {
    return `${fieldName} contains prohibited content`;
  }

  return null;
};

/**
 * Валидация телефона
 */
export const validatePhone = (phone: string): string | null => {
  if (!phone) {
    return null;
  }

  const sanitized = sanitizeString(phone);

  if (sanitized.length > MAX_PHONE_LENGTH) {
    return "Phone number is too long";
  }

  // Разрешаем только цифры, +, -, (, ), пробелы
  if (!/^[\d\s+\-()]+$/.test(sanitized)) {
    return "Phone number contains invalid characters";
  }

  return null;
};

/**
 * Валидация почтового индекса
 */
export const validatePostalCode = (code: string): string | null => {
  if (!code) {
    return null;
  }

  const sanitized = sanitizeString(code);

  if (sanitized.length > MAX_POSTAL_CODE_LENGTH) {
    return "Postal code is too long";
  }

  // Разрешаем цифры, буквы, пробелы, дефисы
  if (!/^[\w\s\-]+$/.test(sanitized)) {
    return "Postal code contains invalid characters";
  }

  return null;
};

